Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift

```diff
@@ -0,0 +1,227 @@
+                            connection: .cancelTimeoutTimer(connection.id)
+                        )
+                    } else {
+                        self.requests.push(request)
+                        return .init(request: .scheduleRequestTimeout(
+                            for: request,
+                            on: preferredEventLoop
+                        ), connection: .none)
+                    }
+                } else {
+                    self.requests.push(request)
+                    let connectionId = self.connections.createNewConnection(on: preferredEventLoop)
+                    return .init(
+                        request: .scheduleRequestTimeout(for: request, on: preferredEventLoop),
+                        connection: .createConnection(connectionId, on: preferredEventLoop)
+                    )
+                }
+            }
+        }
+        
+        mutating func newHTTP2ConnectionEstablished(_ connection: Connection, maxConcurrentStreams: Int) -> Action {
+            self.failedConsecutiveConnectionAttempts = 0
+            self.lastConnectFailure = nil
+            let (index, context) = self.connections.newHTTP2ConnectionEstablished(
+                connection,
+                maxConcurrentStreams: maxConcurrentStreams
+            )
+            return self.nextActionForAvailableConnection(at: index, context: context)
+        }
+        
+        private mutating func nextActionForAvailableConnection(
+            at index: Int,
+            context: HTTP2Connections.AvailableConnectionContext
+        ) -> Action {
+            var requests = self.requests.popFirst(max: context.availableStreams, for: context.eventLoop)
+            let remainingAvailableStreams = context.availableStreams - requests.count
+            // use the remaining available streams for requests without a required event loop
+            requests += self.requests.popFirst(max: remainingAvailableStreams, for: nil)
+            let connection = self.connections.leaseStreams(at: index, count: requests.count)
+            
+            let shouldScheduleIdleTimer = context.isIdle && requests.isEmpty
+            return .init(
+                request: .executeRequestsAndCancelTimeouts(requests, connection),
+                connection: shouldScheduleIdleTimer ? .scheduleTimeoutTimer(connection.id, on: context.eventLoop) : .none
+            )
+        }
+        
+        mutating func newHTTP2MaxConcurrentStreamsReceived(_ connectionID: Connection.ID, newMaxStreams: Int) -> Action {
+            let (index, context) = self.connections.newHTTP2MaxConcurrentStreamsReceived(connectionID, newMaxStreams: newMaxStreams)
+            return nextActionForAvailableConnection(at: index, context: context)
+        }
+        
+        mutating func http2ConnectionGoAwayReceived(_ connectionID: Connection.ID) -> Action {
+            connections.goAwayReceived(connectionID)
```

Comment:
Here we need to check, if we have open requests for the connections el. If we do, we should create a new http2 connection right away.